usr/lib/pkcs11/ep11_stdll/ep11_specific.c

```diff
@@ -2792,22 +2804,140 @@ CK_RV ep11tok_generate_key(STDLL_TokData_t * tokdata, SESSION * session,
+        break;
+    case CKM_SHA512_224:
+        use_libica = ep11_data->libica.ica_sha512_224 != NULL;
+        break;
+    case CKM_SHA512_256:
+        use_libica = ep11_data->libica.ica_sha512_256 != NULL;
+        break;
+    default:
+        use_libica = 0;
+    }
+
+    if (use_libica == 0)
+        TRACE_DEVEL("%s mech=%s is not supported by libica\n", __func__,
+                    ep11_get_ckm(mech));
+
+    return use_libica ? CK_TRUE : CK_FALSE;
+}
+
+CK_RV ep11tok_libica_digest(ep11_private_data_t *ep11_data,
+                            CK_MECHANISM_TYPE mech, libica_sha_context_t *ctx,
+                            CK_BYTE *in_data, CK_ULONG in_data_len,
+                            CK_BYTE *out_data, CK_ULONG *out_data_len,
+                            unsigned int message_part)
+{
+    CK_ULONG hsize;
+    CK_RV rc;
+
+    rc = get_sha_size(mech, &hsize);
+    if (rc != CKR_OK)
+        return rc;
+
+    if (*out_data_len < hsize)
+        return CKR_BUFFER_TOO_SMALL;
+
+    TRACE_DEVEL("%s mech=%s part=%u\n", __func__,
+                ep11_get_ckm(mech), message_part);
+
+    switch (mech) {
+    case CKM_SHA_1:
+        rc = ep11_data->libica.ica_sha1(message_part, in_data_len, in_data,
+                                        &ctx->ctx.sha1, out_data);
+        break;
+    case CKM_SHA224:
+        rc = ep11_data->libica.ica_sha224(message_part, in_data_len, in_data,
+                                          &ctx->ctx.sha256, out_data);
+        break;
+    case CKM_SHA256:
+        rc = ep11_data->libica.ica_sha256(message_part, in_data_len, in_data,
+                                          &ctx->ctx.sha256, out_data);
+        break;
+    case CKM_SHA384:
+        rc = ep11_data->libica.ica_sha384(message_part, in_data_len, in_data,
+                                          &ctx->ctx.sha512, out_data);
+        break;
+    case CKM_SHA512:
+        rc = ep11_data->libica.ica_sha512(message_part, in_data_len, in_data,
+                                          &ctx->ctx.sha512, out_data);
+        break;
+    case CKM_SHA512_224:
+        rc = ep11_data->libica.ica_sha512_224(message_part, in_data_len, in_data,
+                                              &ctx->ctx.sha512, out_data);
+        break;
+    case CKM_SHA512_256:
+        rc = ep11_data->libica.ica_sha512_256(message_part, in_data_len, in_data,
+                                              &ctx->ctx.sha512, out_data);
+        break;
+    default:
+        TRACE_ERROR("%s Invalid mechanism: mech=%s\n", __func__,
+                    ep11_get_ckm(mech));
+        return CKR_MECHANISM_INVALID;
+    }
+
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Libica SHA failed. mech=%s rc=0x%lx\n", __func__,
+                    ep11_get_ckm(mech), rc);
+
+        switch (rc) {
+        case EINVAL:
+            return CKR_ARGUMENTS_BAD;
+        case ENODEV:
+            return CKR_DEVICE_ERROR;
+        default:
+            return CKR_FUNCTION_FAILED;
```

Comment:
i dont like all the ock internal functions returning the pkcs11 CKR_ retcodes: makes it harder to decide if you may change the return codes or if the return code behavior is mandated by pkcs11. Since this is dont core too much at this time.